Adds decoding of strings and string-like blobs based on the charset returned by MySQL

bbeaudreault opened the pull request:
This is the final big PR @harshit-gangal and @ashudeep-sharma.  Almost half of the lines are tests though. As before, this takes some inspiration from mysql-connector-j.

CharsetMapping.java, Collation.java, and MysqlCharset.java come directly from mysql-connector-j. I removed some functions that we didn't use, but otherwise untouched.

Nearly 100% test coverage

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/youtube/vitess/2548)
<!-- Reviewable:end -->

bbeaudreault commented:
The one test failure here is in the go unit tests and seems transient, so is unrelated.

bbeaudreault commented:
@harshit-gangal @ashudeep-sharma any chance you can get some time to look at this this week? It's the last big PR to support charsets appropriately in the Vitess driver.

ashudeep-sharma commented:
@bbeaudreault: we were a bit busy last week with some internal issues, will definitely take a look in this week.

bbeaudreault commented:
Sounds good @ashudeep-sharma, thanks for the update!

harshit-gangal commented on java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java:
```diff
@@ -229,6 +324,34 @@ boolean isReadOnly() throws SQLException {
+    }
+
+
+    public synchronized int getMaxBytesPerCharacter() throws SQLException {
```
SQLException is redundant here.

harshit-gangal commented on java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessConnection.java:
```diff
@@ -842,4 +843,27 @@ public Context createContext(long deadlineAfter) {
         return this.vitessJDBCUrl.getUsername();
     }
+
+    public String getEncodingForIndex(int charsetIndex) throws SQLException {
```
SQLException is redundant here.

harshit-gangal commented on java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessResultSet.java:
```diff
@@ -549,7 +555,50 @@ public Object getObject(int columnIndex) throws SQLException {
+                if (!field.isSingleBit()) {
+                    return bytes;
+                }
+                return byteArrayToBoolean(bytes);
```
byteArrayToBoolean method has redundant SQLExpection.

harshit-gangal commented on java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessResultSetMetaData.java:
```diff
@@ -83,7 +99,11 @@ public boolean isSigned(int column) throws SQLException {
+        if (!field.getConnection().isIncludeAllFields()) {
+            return 0;
+        }
+        return field.getColumnLength() / field.getMaxBytesPerCharacter();
```
Possibility of Divide by Zero error.

harshit-gangal commented on java/jdbc/src/main/java/com/flipkart/vitess/util/charset/Collation.java:
```diff
@@ -0,0 +1,33 @@
+class Collation {
+    public final int index;
+    public final String collationName;
+    public final int priority;
```
Is priority used?

harshit-gangal replied:
Also, why it is declared as public?

bbeaudreault replied:
I tried not to modify the 3 classes in the charset package too much. They were taken directly from the latest version of mysql-connector-j. If you'd like I can make some changes.

harshit-gangal commented on java/jdbc/src/main/java/com/flipkart/vitess/util/charset/MysqlCharset.java:
```diff
@@ -0,0 +1,118 @@
+        this.mblen = mblen;
+        this.priority = priority;
+
+        for (int i = 0; i < javaEncodings.length; i++) {
```
use foreach.

harshit-gangal commented on java/jdbc/src/main/java/com/flipkart/vitess/util/charset/MysqlCharset.java:
```diff
@@ -0,0 +1,118 @@
+class MysqlCharset {
+    public final String charsetName;
+    public final int mblen;
+    public final int priority;
```
Reason for declaring them public? Also, where are they used?

ashudeep-sharma commented:
Went through the first iteration of the review. Overall the changes looks good,  just a few minor comments. Can you please incorporate those. Also as of now i have ignored the changes that you have directly pulled from mysql connector, but i believe there are enough tests written to cover on those aspects.

---

Reviewed 11 of 15 files at r1.
Review status: 11 of 15 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java, line 125 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-KdvU4g5Y_mQOmVNp3t5:-KdvU4g5Y_mQOmVNp3t6:b-i6qm6m) ([raw file](https://github.com/youtube/vitess/blob/d2b2639ccb9e56c69fc043248b15fbce5f276a0f/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java#L125)):*
> ```Java
>         this.characterEncodingAsString = this.characterEncoding.getValueAsString();
>         if (characterEncodingAsString != null) {
>             // Attempt to use the encoding, and bail out if it can't be used
> ```

this code can be abstracted out as a separate function like supportsCharacterEncoding?

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java, line 48 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-KdvVcRVauGgLNdKU5aB:-KdvVcRVauGgLNdKU5aC:b-qkvdis) ([raw file](https://github.com/youtube/vitess/blob/d2b2639ccb9e56c69fc043248b15fbce5f276a0f/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java#L48)):*
> ```Java
>         // fields included when includeFields=IncludedFields.ALL
>         if (connection != null && connection.isIncludeAllFields()) {
>             this.isImplicitTempTable = field.getTable().length() > 5 && field.getTable().startsWith("#sql_");
> ```

Can we define "#sql_" string as implicitTablePattern and mention the reasoning there, will be great if you can add some comment behind the reasoning here like all the tempTables starts with #sql_ pattern etc.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java, line 88 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-KdvX8gK-1PkiMBlIw3J:-KdvX8gLh7EWM6ed_tgW:b-4pleaa) ([raw file](https://github.com/youtube/vitess/blob/d2b2639ccb9e56c69fc043248b15fbce5f276a0f/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java#L88)):*
> ```Java
>                     this.encoding = "UTF-16";
>                 }
>                 // MySQL encodes JSON data with utf8mb4.
> ```

Are these two if conditions orthogonal? If yes then Query.Type.JSON can always override? Need some more details in here.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java, line 102 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-KdvXXh-k0vy68ryPnEV:-KdvXXh-k0vy68ryPnEW:b-mzm7fp) ([raw file](https://github.com/youtube/vitess/blob/d2b2639ccb9e56c69fc043248b15fbce5f276a0f/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java#L102)):*
> ```Java
>             } else {
>                 // Default encoding for number-types and date-types
>                 this.encoding = "US-ASCII";
> ```

What will be this.javaType value in this else condition?

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessResultSet.java, line 760 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-Kdva7YjsfbdVvcc-M09:-Kdva7YjsfbdVvcc-M0A:b-2fk95h) ([raw file](https://github.com/youtube/vitess/blob/d2b2639ccb9e56c69fc043248b15fbce5f276a0f/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessResultSet.java#L760)):*
> ```Java
>     }
> 
>     //Unsupported Methods
> ```

Please move it below, i can see these methods are being used. we can move it before getAsciiStream() method.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/util/MysqlDefs.java, line 76 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-KdvbkspbRTN3JJAfzv5:-KdvbksqDYllJKmona_9:bpsy0di) ([raw file](https://github.com/youtube/vitess/blob/d2b2639ccb9e56c69fc043248b15fbce5f276a0f/java/jdbc/src/main/java/com/flipkart/vitess/util/MysqlDefs.java#L76)):*
> ```Java
>     static final int INIT_DB = 2;
>     public static final int LENGTH_BLOB = 65535;
>     public static final int LENGTH_LONGBLOB = Integer.MAX_VALUE;
> ```

Any reason why LENGTH_LONGBLOB type being changed to int from long, I see in mysql jdbc it still holds the value 4294967295L. Ref : https://github.com/mysql/mysql-connector-j/blob/402933ef52cad9aa82624e80acbea46e3a701ce6/src/com/mysql/jdbc/MysqlDefs.java#L136

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2548#-:-KdvdzVA2pcOw3CmEso6:b-9oli0j)*
<!-- Sent from Reviewable.io -->

bbeaudreault commented:
Thanks guys, pushed some changes to respond to your comments. Did not make changes to the `charsets` package, to keep the diff minimal compared to mysql-connector-j

bbeaudreault commented:
Review status: 6 of 15 files reviewed at latest revision, 13 unresolved discussions.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java, line 125 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-KdvU4g5Y_mQOmVNp3t5:-Ke-os4in2_6KlpQA7v4:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/2ee53b491025aa6c17af614ac43e9b37153731e3/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java#L125)):*
<details><summary><i>Previously, ashudeep-sharma (Ashu) wrote…</i></summary><blockquote>

this code can be abstracted out as a separate function like supportsCharacterEncoding?
</blockquote></details>

Done.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java, line 48 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-KdvVcRVauGgLNdKU5aB:-Ke-ot-NZyxEdPcvk2mN:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/2ee53b491025aa6c17af614ac43e9b37153731e3/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java#L48)):*
<details><summary><i>Previously, ashudeep-sharma (Ashu) wrote…</i></summary><blockquote>

Can we define "#sql_" string as implicitTablePattern and mention the reasoning there, will be great if you can add some comment behind the reasoning here like all the tempTables starts with #sql_ pattern etc.
</blockquote></details>

Done.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java, line 88 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-KdvX8gK-1PkiMBlIw3J:-Ke-otoJ22WWz79D5QpT:b-l5qxkn) ([raw file](https://github.com/youtube/vitess/blob/2ee53b491025aa6c17af614ac43e9b37153731e3/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java#L88)):*
<details><summary><i>Previously, ashudeep-sharma (Ashu) wrote…</i></summary><blockquote>

Are these two if conditions orthogonal? If yes then Query.Type.JSON can always override? Need some more details in here.
</blockquote></details>

Yes the two are orthogonal. We're just doing some fixup after getting the passed encoding. I add a few comments to clarify. This logic was adapted from mysql-connector-j.  I simplified a lot of it to account for our only supporting 5.6+ and such, but did not want to totally eliminate these clauses since they seemed legit.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java, line 102 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-KdvXXh-k0vy68ryPnEV:-Ke-pDwBRA2-uD3TR8ca:bjx83xv) ([raw file](https://github.com/youtube/vitess/blob/2ee53b491025aa6c17af614ac43e9b37153731e3/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java#L102)):*
<details><summary><i>Previously, ashudeep-sharma (Ashu) wrote…</i></summary><blockquote>

What will be this.javaType value in this else condition?
</blockquote></details>

Clarified. We already set the javaType here so no need to re-set

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java, line 344 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-Kdk3a-jKewh2BMaPSOB-r1-344:-Ke-pGbVGsv8aQW0jjiC:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/2ee53b491025aa6c17af614ac43e9b37153731e3/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java#L344)):*
<details><summary><i>Previously, harshit-gangal wrote…</i></summary><blockquote>

SQLException is redundant here.
</blockquote></details>

Done.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessConnection.java, line 847 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-Kdk3a-jKewh2BMaPSOC-r1-847:-Ke-pMzHvD0T8zw3iphb:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/2ee53b491025aa6c17af614ac43e9b37153731e3/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessConnection.java#L847)):*
<details><summary><i>Previously, harshit-gangal wrote…</i></summary><blockquote>

SQLException is redundant here.
</blockquote></details>

Done.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessResultSet.java, line 575 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-Kdk3a-jKewh2BMaPSOD-r1-575:-Ke-pON5c9fJdo-dCTB9:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/2ee53b491025aa6c17af614ac43e9b37153731e3/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessResultSet.java#L575)):*
<details><summary><i>Previously, harshit-gangal wrote…</i></summary><blockquote>

byteArrayToBoolean method has redundant SQLExpection.
</blockquote></details>

Done.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessResultSet.java, line 760 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-Kdva7YjsfbdVvcc-M09:-Ke-pPAK5rnw1_To6y7w:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/2ee53b491025aa6c17af614ac43e9b37153731e3/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessResultSet.java#L760)):*
<details><summary><i>Previously, ashudeep-sharma (Ashu) wrote…</i></summary><blockquote>

Please move it below, i can see these methods are being used. we can move it before getAsciiStream() method.
</blockquote></details>

Done.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessResultSetMetaData.java, line 106 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-Kdk3a-jKewh2BMaPSOE-r1-106:-Ke-pPtmYor1bopflIuc:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/2ee53b491025aa6c17af614ac43e9b37153731e3/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/VitessResultSetMetaData.java#L106)):*
<details><summary><i>Previously, harshit-gangal wrote…</i></summary><blockquote>

Possibility of Divide by Zero error.
</blockquote></details>

Done.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/util/MysqlDefs.java, line 76 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-KdvbkspbRTN3JJAfzv5:-Ke-pQeN6DNy3fHisypT:bbwy0gr) ([raw file](https://github.com/youtube/vitess/blob/2ee53b491025aa6c17af614ac43e9b37153731e3/java/jdbc/src/main/java/com/flipkart/vitess/util/MysqlDefs.java#L76)):*
<details><summary><i>Previously, ashudeep-sharma (Ashu) wrote…</i></summary><blockquote>

Any reason why LENGTH_LONGBLOB type being changed to int from long, I see in mysql jdbc it still holds the value 4294967295L. Ref : https://github.com/mysql/mysql-connector-j/blob/402933ef52cad9aa82624e80acbea46e3a701ce6/src/com/mysql/jdbc/MysqlDefs.java#L136
</blockquote></details>

This is here because @alainjobart changed VTGate to return int instead of long for the FieldLength metadata. In the mysql-connector-j they had a function:

```
    private static int clampedGetLength(Field f) {
        long fieldLength = f.getLength();

        if (fieldLength > Integer.MAX_VALUE) {
            fieldLength = Integer.MAX_VALUE;
        }

        return (int) fieldLength;
    }
```

We don't need that because f.getLength() is int for us. But I got some type conversion errors in the comparisons we do make, so I changed these to integer to reflect reality.  I'll add a comment.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/util/charset/MysqlCharset.java, line 17 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-Kdk3a-jKewh2BMaPSOJ-r1-17:-Ke-qKt1WSvDfUk2joUM:bifrsd2) ([raw file](https://github.com/youtube/vitess/blob/2ee53b491025aa6c17af614ac43e9b37153731e3/java/jdbc/src/main/java/com/flipkart/vitess/util/charset/MysqlCharset.java#L17)):*
<details><summary><i>Previously, harshit-gangal wrote…</i></summary><blockquote>

Reason for declaring them public? Also, where are they used?
</blockquote></details>

See above comment about not modifying these to keep the diff minimal

---

*[java/jdbc/src/main/java/com/flipkart/vitess/util/charset/MysqlCharset.java, line 38 at r1](https://reviewable.io:443/reviews/youtube/vitess/2548#-Kdk3a-jKewh2BMaPSOJ-r1-38:-Ke-qONQqSVj45lpCs6t:bifrsd2) ([raw file](https://github.com/youtube/vitess/blob/2ee53b491025aa6c17af614ac43e9b37153731e3/java/jdbc/src/main/java/com/flipkart/vitess/util/charset/MysqlCharset.java#L38)):*
<details><summary><i>Previously, harshit-gangal wrote…</i></summary><blockquote>

use foreach.
</blockquote></details>

See above comment about not modifying these to keep the diff minimal

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2548)*
<!-- Sent from Reviewable.io -->

bbeaudreault commented:
Ready for re-review

---

Review status: 6 of 15 files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2548#-:-Ke-qbgTQD9gHwlbX-Nx:b-6drotj)*
<!-- Sent from Reviewable.io -->

bbeaudreault commented:
@ashudeep-sharma @harshit-gangal if I can get a re-review when you get a chance, that'd be great. I don't want to pressure, but I'll be leaving for 2 weeks on Saturday. Would be great to get merged before then, if possible.

ashudeep-sharma commented:
Some minor Comments. Rests Looks good to me. The only concern is here we are directly pulling out some code from mysql.Should we actually understand the details about code. Let me know your thoughts @harshit-gangal , @bbeaudreault .

---

Reviewed 4 of 15 files at r1, 4 of 5 files at r2.
Review status: 13 of 15 files reviewed at latest revision, 20 unresolved discussions.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java, line 129 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke7hK8UoAignPSlj4Fj:-Ke7hK8UoAignPSlj4Fk:b-nyadkl) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java#L129)):*
> ```Java
>     private void checkConfiguredEncodingSupport() throws SQLException {
>         if (characterEncodingAsString != null) {
>             // Attempt to use the encoding, and bail out if it can't be used
> ```

Please move the comment on top of function.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java, line 143 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke7hODzEb1xgjNliOz3:-Ke7gpOYlFNbGR0i3bOQ:b-aux23k) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java#L143)):*
> ```Java
>     }
> 
>     protected DriverPropertyInfo[] exposeAsDriverPropertyInfoInternal(Properties info, int slotsToReserve) throws SQLException {
> ```

Any reason for making this function protected, i can see its only being used in this file only.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java, line 89 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke7nrEJXb9a4tY1A7Ry:-Ke7nrEK6fCv9vU-xbhn:bdluisr) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java#L89)):*
> ```Java
>                 // ucs2, utf16, and utf32 cannot be used as a client character set, but if it was received from server
>                 // under some circumstances we can parse them as utf16
>                 if ("UnicodeBig".equals(this.encoding)) {
> ```

can the comparison be case sensitive here? or are we specifically looking for "UnicodeBig" string?

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java, line 425 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke7hxlvjnGqO0EXNTmh:-Ke7hxlvjnGqO0EXNTmi:b-vlw7u2) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java#L425)):*
> ```Java
>     }
> 
>     public Query.Type getVitessType() {
> ```

This function can be made private, its only used in this file.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/util/StringUtils.java, line 388 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke8-KEgQwQL9F31fnJh:-Ke8-KEgQwQL9F31fnJi:bc1f4b) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/util/StringUtils.java#L388)):*
> ```Java
>     public static String toString(byte[] value, int offset, int length, String encoding) throws UnsupportedEncodingException {
>         Charset cs = findCharset(encoding);
> ```

Remove new line

---

*[java/jdbc/src/main/java/com/flipkart/vitess/util/StringUtils.java, line 428 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke8-P_PO0F5iAoJ-hAP:-Ke8-P_PO0F5iAoJ-hAQ:b-w1non7) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/util/StringUtils.java#L428)):*
> ```Java
>     public static byte[] getBytes(String value, int offset, int length, String encoding) throws UnsupportedEncodingException {
>         Charset cs = findCharset(encoding);
> ```

Remove newLine

---

*[java/jdbc/src/main/java/com/flipkart/vitess/util/StringUtils.java, line 459 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke8-YDrI7NfOiMxINbB:-Ke8-YDsq0kBtuMGbejJ:b-psygq0) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/util/StringUtils.java#L459)):*
> ```Java
>         }
>     }
> }
> ```

There are some extra lines in this file. Please remove it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2548#-:-Ke8-wRajwIsSYIFBAsq:beu4rda)*
<!-- Sent from Reviewable.io -->

bbeaudreault commented:
Thanks ashu. I'll take care of those today.

In terms of understanding the charsets package code, I don't think the code is too complicated. I have looked at it a bit and do understand it. I was avoiding making changes to it incase in the future we want to port more improvements/changes to those classes as they are made in mysql-connector-j. Limiting the diff will make that easier.

The most important part of that package is CharsetMapping, which is a long list of hardcoded mappings from mysql charsets/collations to java charsets. The list is indexed into a couple maps, and functions are provided for reading them. The remaining two classes are just used to support CharsetMapping.  Some of the seemingly unused fields that @harshit-gangal pointed out are because I deleted some functions from CharsetMapping that we have no need for, to simplify things. Otherwise the class is pretty long.

If we all decide to I'd be happy to go through and simplify those classes a bit further, like removing the unused parts and documenting a bit.  Testing them is a bit weird though since it's just a static list of encodings. We'd need to test each of them individually and it'd be brittle. I agree that the rest of the tests seem enough.

harshit-gangal commented:
LGTM. Need to address few comments from @ashudeep-sharma

---

Reviewed 3 of 15 files at r1, 1 of 5 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2548#-:-Ke9t5xFTWFswNeLlzqm:bxpsutc)*
<!-- Sent from Reviewable.io -->

ashudeep-sharma commented:
@bbeaudreault Lets incorporate the comments and merge it for now, we can take the simplification in next PR. The only other thing that we need to keep an eye on if something gets changes in mysql connector to any of these files.We have to port those changes here as well.

bbeaudreault commented:
Review status: 12 of 15 files reviewed at latest revision, 13 unresolved discussions.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java, line 129 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke7hK8UoAignPSlj4Fj:-KeF-WeiBeucf140qv3c:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java#L129)):*
<details><summary><i>Previously, ashudeep-sharma (Ashu) wrote…</i></summary><blockquote>

Please move the comment on top of function.
</blockquote></details>

Done.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java, line 143 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke7hODzEb1xgjNliOz3:-KeF-_ZgOGHhc6UXZrhE:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/ConnectionProperties.java#L143)):*
<details><summary><i>Previously, ashudeep-sharma (Ashu) wrote…</i></summary><blockquote>

Any reason for making this function protected, i can see its only being used in this file only.
</blockquote></details>

Done.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java, line 89 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke7nrEJXb9a4tY1A7Ry:-KeF08ZTPx4a_f-sQdW5:buazr9o) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java#L89)):*
<details><summary><i>Previously, ashudeep-sharma (Ashu) wrote…</i></summary><blockquote>

can the comparison be case sensitive here? or are we specifically looking for "UnicodeBig" string?
</blockquote></details>

UnicodeBig is the actual name of the encoding. I think it should be case-sensitive

---

*[java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java, line 425 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke7hxlvjnGqO0EXNTmh:-KeF0gF-pDmwuU1jXs9Q:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/jdbc/FieldWithMetadata.java#L425)):*
<details><summary><i>Previously, ashudeep-sharma (Ashu) wrote…</i></summary><blockquote>

This function can be made private, its only used in this file.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2548)*
<!-- Sent from Reviewable.io -->

bbeaudreault commented:
Addressed all comments. Thanks guys

---

Review status: 12 of 15 files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2548#-:-KeF1fch9gzedLfakTFJ:b-qpq95k)*
<!-- Sent from Reviewable.io -->

bbeaudreault commented:
Review status: 12 of 15 files reviewed at latest revision, 13 unresolved discussions.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/util/StringUtils.java, line 388 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke8-KEgQwQL9F31fnJh:-KeF1tb2mhQCN0S2-tVE:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/util/StringUtils.java#L388)):*
<details><summary><i>Previously, ashudeep-sharma (Ashu) wrote…</i></summary><blockquote>

Remove new line
</blockquote></details>

Done.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/util/StringUtils.java, line 428 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke8-P_PO0F5iAoJ-hAP:-KeF1u0K5OsYvo7UOZNC:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/util/StringUtils.java#L428)):*
<details><summary><i>Previously, ashudeep-sharma (Ashu) wrote…</i></summary><blockquote>

Remove newLine
</blockquote></details>

Done.

---

*[java/jdbc/src/main/java/com/flipkart/vitess/util/StringUtils.java, line 459 at r3](https://reviewable.io:443/reviews/youtube/vitess/2548#-Ke8-YDrI7NfOiMxINbB:-KeF1uXUl96Mw3wZENNk:b-896fix) ([raw file](https://github.com/youtube/vitess/blob/857089de7cce9745c172444bdd4a8846f9c65900/java/jdbc/src/main/java/com/flipkart/vitess/util/StringUtils.java#L459)):*
<details><summary><i>Previously, ashudeep-sharma (Ashu) wrote…</i></summary><blockquote>

There are some extra lines in this file. Please remove it.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2548)*
<!-- Sent from Reviewable.io -->

ashudeep-sharma commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/youtube/vitess/2548#-:-KeF454miVe7sZ9kxpu0:bnfp4nl)*
<!-- Sent from Reviewable.io -->

michael-berlin commented:
LGTM as the code owners LGTM'd it.
